Signal/src/view controllers/MessageComposeTableViewController.m

```diff
@@ -24,24 +20,44 @@ @interface MessageComposeTableViewController () <UISearchBarDelegate,
                                                  UISearchResultsUpdating,
                                                  MFMessageComposeViewControllerDelegate>
 
-@property (nonatomic, strong) IBOutlet UITableViewCell *inviteCell;
-@property (nonatomic, strong) IBOutlet OWSNoSignalContactsView *noSignalContactsView;
+@property (nonatomic) IBOutlet UITableViewCell *inviteCell;
+@property (nonatomic) UITableViewCell *conversationForNonContactCell;
+@property (nonatomic) UITableViewCell *inviteViaSMSCell;
+@property (nonatomic) IBOutlet OWSNoSignalContactsView *noSignalContactsView;
 
-@property (nonatomic) UIButton *sendTextButton;
-@property (nonatomic, strong) UISearchController *searchController;
-@property (nonatomic, strong) UIActivityIndicatorView *activityIndicator;
-@property (nonatomic, strong) UIBarButtonItem *addGroup;
-@property (nonatomic, strong) UIView *loadingBackgroundView;
+@property (nonatomic) UISearchController *searchController;
+@property (nonatomic) UIActivityIndicatorView *activityIndicator;
+@property (nonatomic) UIBarButtonItem *addGroup;
+@property (nonatomic) UIView *loadingBackgroundView;
 
 @property (nonatomic) NSString *currentSearchTerm;
 @property (copy) NSArray<Contact *> *contacts;
 @property (copy) NSArray<Contact *> *searchResults;
 @property (nonatomic, readonly) OWSContactsManager *contactsManager;
 
+// This property should be set IFF the current search text can
+// be parsed as a phone number. If set, it contains a E164 value.
+@property (nonatomic) NSString *searchPhoneNumber;
+// This dictionary is used to cache the set of phone numbers
+// which are known to correspond to Signal accounts.
+@property (nonatomic) NSMutableSet *phoneNumberAccountSet;
+
+@property (nonatomic) BOOL isBackgroundViewHidden;
+
 @end
 
-NSInteger const MessageComposeTableViewControllerSectionInvite = 0;
-NSInteger const MessageComposeTableViewControllerSectionContacts = 1;
+// The "special" sections are used to display (at most) one of three cells:
+//
+// * "New conversation for non-contact" if user has entered a phone
+//    number which corresponds to a signal account, or:
+// * "Send invite via SMS" if user has entered a phone number
+//    which is not known to correspond to a signal account, or:
+// * "Invite contacts" if the invite flow is available, or:
+// * Nothing, otherwise.
+NSInteger const MessageComposeTableViewControllerSectionInviteNonContactConversation = 0;
```

Comment:
Yeah, I do too - I was just trying to refactor this as little as possible. Anyway, done.